src/mechanics/dofs/DofNumbering.cpp

```diff
@@ -0,0 +1,95 @@
+#include "mechanics/dofs/DofNumbering.h"
+
+using namespace NuTo;
+
+//! @brief build dof numbering, starting at 0, for all `nodes` regardless of constraints
+//! @return total number of dofs in `nodes`
+int InitialUnconstrainedNumbering(const Groups::Group<NodeSimple>& nodes)
+{
+    int dofNumber = 0;
+    for (auto& node : nodes)
+        for (int iComponent = 0; iComponent < node.GetNumValues(); ++iComponent)
+            node.SetDofNumber(iComponent, dofNumber++);
+    return dofNumber;
+}
+
+constexpr int INDEPENDENT = -1;
+
+//! @brief Find the status (INDEPENDENT or dependent) for all dof numbers.
+//! @return The status vector contains two information. Value INDEPENDENT means, well..., independent. Any value other
+//! than that is the index of the equation where this dof is the dependent one.
+std::vector<int> GetStatusOfDofNumber(const ConstraintPde::Constraints& constraints, DofType dof, int numDofs)
+{
+    std::vector<int> dofStatus(numDofs, INDEPENDENT); // initialize as INDEPENDENT
+    for (int iEquation = 0; iEquation < constraints.GetNumEquations(dof); ++iEquation)
+    {
+        int dependentDofNumber = constraints.GetEquation(dof, iEquation).GetDependentDofNumber();
+
+        // FIRST CHECK: Are there multiple dependent dofs
+        if (dofStatus[dependentDofNumber] != INDEPENDENT)
+            throw NuTo::Exception(
+                    __PRETTY_FUNCTION__,
+                    "Dof number " + std::to_string(dependentDofNumber) +
+                            " already constrained. You are only allowed to have a dependent dof in one equation.");
+
+        // add the equation index to the status.
+        dofStatus[dependentDofNumber] = iEquation;
+    }
+
+    // go through all terms that are not dependent and check if they contain dependent dofs
+    for (int iEquation = 0; iEquation < constraints.GetNumEquations(dof); ++iEquation)
```

Comment:
Would it be nice to have such a check when a constraint is created or a term is added?

Reply:
That is a _very_ good point that bothers me as well. 

This method is called when all dofs have a unique, arbitrary number. So we can do all these checks based on this number. 
If we want to do it earlier, and it _definitely makes perfect sense_, we do not have a unique dof numbering. Thus, we have do deal with something like ´std::pair<Node, component>` do uniquely identify the dof. I hope @vhirtham and I can think of a clean solution.